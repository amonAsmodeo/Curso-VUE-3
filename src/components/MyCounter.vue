<template>
  <section>
    <h3>Counter: {{ counter }}</h3>
    <h3 data-testid="square-label">Square: {{ squareCounter }}</h3>

    <div>
      <button class="p-2 m-1 bg-red-500 text-2xl hover:bg-red-800" @click="counter++">+1</button>
      <button class="p-2 m-1 bg-blue-500 text-2xl hover:bg-blue-800" @click="counter--">-1</button>
    </div>
  </section>
</template>

<script lang="ts" setup>
import { useCounter } from '@/composables/useCounter';

// /* en ts */
interface Props {
  value?: number /* con el signo de interrogacion dice que es opcional */;
}

const props = defineProps<Props>();

console.log('las props en MyCounter.vue: ', props);

const { counter, squareCounter } = useCounter(props.value);

// import { computed, ref } from 'vue';

/*forma en JS */
// const props = defineProps({
//   value: { type: Number, required: true },
// });

/* indica que si el valor de props es undefined el valor por defecto sera 5*/
// const counter = ref(props.value ?? 5);
// const SquareCounter = computed(() => counter.value * counter.value);

// const counterAdd = () => {
//   return (counter.value = counter.value + 1);
// };

// const counterRes = () => {
//   return (counter.value = counter.value - 1);
// };
</script>
