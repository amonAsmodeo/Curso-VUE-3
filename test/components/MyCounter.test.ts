import MyCounter from '@/components/MyCounter.vue';
import { describe, expect, test } from 'vitest';
import { mount } from '@vue/test-utils';

describe('<Pruebas de MyCounter />', () => {
  test('sould match snapshot', () => {
    // Preparar la pruena
    const wrapper = mount(MyCounter, {
      props: {
        value: 5,
      },
    });

    // prueba que la vista se renderice igual a la captura que se tenia previamente
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Render the counter value correctly', () => {
    const value = 5;

    const wrapper = mount(MyCounter, {
      props: {
        value: value,
      },
    });

    expect(wrapper.find('h3').text()).toContain(`Counter: ${value}`);
    expect(wrapper.find('[data-testid="square-label"]').text()).toContain(
      `Square: ${value * value}`,
    );
  });

  test('increments the couner when +1 button is clicked', async () => {
    const value = 5;

    const wrapper = mount(MyCounter, {
      props: {
        value: value,
      },
    });

    // obtengo todos los h3 y desestructuro
    const [counterLabel, squareLabel] = wrapper.findAll('h3');
    // obtenemos el primer boton
    const btnIncrement = wrapper.find('button');
    // se simula click en el boton
    await btnIncrement.trigger('click');

    expect(counterLabel.text()).toContain(`Counter: ${value + 1}`);
    expect(squareLabel.text()).toContain(`Square: ${(value + 1) * (value + 1)}`);
  });

  test('decrements the counter when -1 button is clicked twice', async () => {
    const value = 5;

    const wrapper = mount(MyCounter, {
      props: {
        value: value,
      },
    });

    // obtener todos los h3 y desestructuro los elementos
    const [counterLabel, squareLabel] = wrapper.findAll('h3');

    // obtener el segundo boton por desestructuracion,
    // el primer elemento no interesa por el momento por eso no hay un nombre para el elemento
    const [, btnDecrement] = wrapper.findAll('button');

    //  simular 2 clicks en el boton para restar
    await btnDecrement.trigger('click');
    await btnDecrement.trigger('click');

    expect(counterLabel.text()).toContain(`Counter: ${value - 2}`);
    expect(squareLabel.text()).toContain(`Square: ${(value - 2) * (value - 2)}`);
  });
});
