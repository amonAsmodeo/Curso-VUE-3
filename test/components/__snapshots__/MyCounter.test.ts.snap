// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Pruebas de MyCounter /> > sould match snapshot 1`] = `
"<section>
  <h3>Counter: 5</h3>
  <h3 data-testid="square-label">Square: 25</h3>
  <div><button class="p-2 m-1 bg-red-500 text-2xl hover:bg-red-800">+1</button><button class="p-2 m-1 bg-blue-500 text-2xl hover:bg-blue-800">-1</button></div>
</section>"
`;
